Init write-cache asynchronously

carpawell opened the pull request.

fyrchik commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ Changelog for FrostFS Node
 
 ### Changed
 - `common.PrintVerbose` prints via `cobra.Command.Printf` (#1962)
+-  Write-cache inits asynchronously (#32) 
```
`Initialize write-cache asynchronously`?

fyrchik commented on pkg/local_object_storage/writecache/init.go:
```diff
@@ -12,11 +13,55 @@ import (
+	if !async {
+		c.fsTreeFlushMarkUpdate()
+		c.dbFlushMarkUpdate()
+		c.inited = true
```
Can you add a comment about why we don't need a mutex here?

carpawell replied:
added but depends on async flag discussion (seems like it could be even removed)

carpawell commented on pkg/local_object_storage/writecache/init.go:
```diff
@@ -12,11 +13,55 @@ import (
+		localWG.Done()
+	}()
+
+	go func() {
```
hw, yeah, would be more reliable, added

carpawell commented on pkg/local_object_storage/writecache/writecache.go:
```diff
@@ -49,6 +49,7 @@ type cache struct {
 	mtx sync.RWMutex
 
 	mode    mode.Mode
+	inited  bool
```
longer but dont mind

carpawell commented on pkg/local_object_storage/writecache/writecache.go:
```diff
@@ -143,8 +144,8 @@ func (c *cache) Open(readOnly bool) error {
 // Init runs necessary services.
-func (c *cache) Init() error {
-	c.initFlushMarks()
+func (c *cache) Init(async bool) error {
```
yes, that's an ugly change. problems appear when you try to (unit)test it. `WriteCache` should either be able to init in sync mode or have a feedback channel to signal about its ready state then. i decided to open PR anyway and discuss that. your thoughts?

acid-ant replied:
vote for new option for writecache `WithAsyncInitialization` and init it in background by default.

fyrchik replied:
Can we just use `require.Eventually` in tests?

carpawell replied:
> Can we just use require.Eventually in tests?

yes, but with magic numbers (how oftern and how long wait for that status (and how? try to `Put` object? export `Status` func?)). do we want it?

> vote for new option for writecache WithAsyncInitialization and init it in background by default.

we have `Init` -> `Close` -> `Init` scenario. i could imagine that sync/async initialization could be a config value in future. also, IMO creation option should not change `Init` behavior

fyrchik replied:
> do we want it?
We already have it in some places, I don't mind having it here.

> i could imagine that sync/async initialization could be a config value in future
Well, it could, but we don't _need_ it now.

> creation option should not change Init behavior
Why so?

carpawell replied:
> I don't mind having it here

ok, dropped `async` flag

Commit:
[#32] node: Init write-cache asynchronously

Signed-off-by: Pavel Karpy <p.karpy@yadro.com>

fyrchik commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ Changelog for FrostFS Node
 ### Changed
 - `common.PrintVerbose` prints via `cobra.Command.Printf` (#1962)
 - Env prefix in configuration changed to `FROSTFS_*` (#43)
+-  Initialize write-cache asynchronously (#32)
```
2 spaces here

fyrchik commented on pkg/local_object_storage/writecache/flush_test.go:
```diff
@@ -316,6 +318,27 @@ func newObject(t *testing.T, size int) (*object.Object, []byte) {
+
+		_, err = wc.Put(prm)
+		return !errors.Is(err, ErrNotInitialized)
+	}, time.Second, 100*time.Millisecond)
```
Is it enough with `-race` flag?

carpawell replied:
it was not but because of another reason, fixed races

fyrchik commented on pkg/local_object_storage/writecache/init.go:
```diff
@@ -13,10 +14,51 @@ import (
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
```
If we use `c.wg` here, do we need to use it in 2 previous goroutines?

fyrchik commented on pkg/local_object_storage/writecache/init.go:
```diff
@@ -13,10 +14,51 @@ import (
 	var prm common.IteratePrm
 	prm.LazyHandler = func(addr oid.Address, _ func() ([]byte, error)) error {
+		select {
+		case <-c.closeCh:
```
It is enough for `setMode` to work too?

carpawell replied:
was not, fixed

fyrchik commented on pkg/local_object_storage/writecache/mode.go:
```diff
@@ -11,6 +11,9 @@ import (
 var ErrReadOnly = logicerr.New("write-cache is in read-only mode")
 
+// ErrNotInitialized is returned when write-cache is initializing.
+var ErrNotInitialized = logicerr.New("write-cache is not initialized")
```
I would add ` yet`, not to spread panic among testers, administrators and service engineers.

fyrchik commented on pkg/local_object_storage/writecache/writecache.go:
```diff
@@ -164,6 +165,8 @@ func (c *cache) Close() error {
 		c.closeCh = nil
 	}
 
+	c.initialized = false
+
```
Is this line necessary?

fyrchik commented on pkg/local_object_storage/writecache/flush_test.go:
```diff
@@ -316,6 +318,27 @@ func newObject(t *testing.T, size int) (*object.Object, []byte) {
+	var err error
+
+	require.Eventually(t, func() bool {
+		var prm common.PutPrm
+		prm.Address = objectCore.AddressOf(obj)
+		prm.Object = obj
+		prm.RawData = data
+
+		_, err = wc.Put(prm)
```
The tests are in the `writecache` package, can we just directly check for `initialized` flag?

Commit:
[#32] node: Always close general components after testing

It will prevent test fails with `-race` flag on components that have
background processes and make some actions on test framework.

Signed-off-by: Pavel Karpy <p.karpy@yadro.com>

carpawell commented:
@fyrchik, fixed tests with `-race` flags but now PR is a little bit more complex. Review one more time, please.

acid-ant commented on pkg/local_object_storage/writecache/generic_test.go:
```diff
@@ -28,3 +28,17 @@ func TestGeneric(t *testing.T) {
 	storagetest.TestAll(t, newCache)
 }
+
+//func TestTest(t *testing.T) {
```
Looks like needs to be deleted.

carpawell replied:
wow, dropped

carpawell commented on pkg/local_object_storage/writecache/flush_test.go:
```diff
@@ -316,6 +317,19 @@ func newObject(t *testing.T, size int) (*object.Object, []byte) {
+	require.Eventually(t, func() bool {
+		rawWc := wc.(*cache)
+
+		rawWc.modeMtx.RLock()
+		defer rawWc.modeMtx.RUnlock()
```
it was not an `atomic` before and then i just changed it and forget other cases to be adopted too

fixed

carpawell commented on pkg/local_object_storage/writecache/mode.go:
```diff
@@ -11,20 +11,37 @@ import (
+		c.initWG.Wait()
+		c.stopInitCh = make(chan struct{})
+
+		defer c.initFlushMarks()
```
flashbacks from the previous implementation

looked at that one more time and it seems like an error-free and non-degraded mode switch is enough for running `initFlushMarks` one more time? check the updated version, please

carpawell commented on pkg/local_object_storage/writecache/mode.go:
```diff
@@ -11,20 +11,37 @@ import (
 	}
 
+	if !c.initialized.Load() {
+		close(c.stopInitCh)
+
+		c.initWG.Wait()
+		c.stopInitCh = make(chan struct{})
```
sure, added comments to that fields
